sklearn/ensemble/_gradient_boosting.pyx

```diff
@@ -84,22 +88,111 @@ cdef void _predict_regression_tree_inplace_fast(DTYPE_t *X,
         shape ``(n_samples, K)``.
     """
     cdef Py_ssize_t i
-    cdef int32 node_id
     cdef Node *node
     for i in range(n_samples):
         node = root_node
         # While node not a leaf
-        while node.left_child != -1 and node.right_child != -1:
+        while node.left_child != TREE_LEAF:
             if X[i * n_features + node.feature] <= node.threshold:
                 node = root_node + node.left_child
             else:
                 node = root_node + node.right_child
         out[i * K + k] += scale * value[node - root_node]
 
+def _predict_regression_tree_stages_sparse(np.ndarray[object, ndim=2] estimators,
+                                           object X, double scale,
+                                           np.ndarray[float64, ndim=2] out):
+    """Predicts output for regression tree inplace and adds scaled value to ``out[i, k]``.
+
+    The function assumes that the ndarray that wraps ``X`` is csr_matrix.
+    """
+    cdef DTYPE_t* X_data = <DTYPE_t*>(
+        <np.ndarray[ndim=1, dtype=DTYPE_t]>X.data).data
+    cdef INT32_t* X_indices = <INT32_t*>(
+        <np.ndarray[ndim=1, dtype=INT32_t]> X.indices).data
+    cdef INT32_t* X_indptr = <INT32_t*>(
+        <np.ndarray[ndim=1, dtype=INT32_t]> X.indptr).data
+
+    cdef SIZE_t n_samples = X.shape[0]
+    cdef SIZE_t n_features = X.shape[1]
+    cdef SIZE_t n_stages = estimators.shape[0]
+    cdef SIZE_t n_outputs = estimators.shape[1]
+
+    # Initialize output
+    cdef float64* out_ptr = <float64*> out.data
+
+    # Indices and temporary variables
+    cdef SIZE_t sample_i = 0
+    cdef SIZE_t feature_i = 0
+    cdef SIZE_t stage_i = 0
+    cdef SIZE_t output_i = 0
+    cdef Node *root_node = NULL
+    cdef Node *node = NULL
+    cdef double *value = NULL
+
+    cdef Tree tree
+    cdef Node** nodes = NULL
+    cdef double** values = NULL
+    safe_realloc(&nodes, n_stages * n_outputs)
+    safe_realloc(&values, n_stages * n_outputs)
+    for stage_i in range(n_stages):
+        for output_i in range(n_outputs):
+            tree = estimators[stage_i, output_i].tree_
+            nodes[stage_i * n_outputs + output_i] = tree.nodes
+            values[stage_i * n_outputs + output_i] = tree.value
+
+    # Initialize auxiliary data-structure
+    cdef DTYPE_t feature_value = 0.
+    cdef DTYPE_t* X_sample = NULL
+
+    # feature_to_sample as a data structure records the last seen sample
+    # for each feature; functionally, it is an efficient way to identify
+    # which features are nonzero in the present sample.
+    cdef SIZE_t* feature_to_sample = NULL
+
+    safe_realloc(&X_sample, n_features)
```

Comment:
Yes, it's from _apply_sparse_csr.
@jmschrei, you are right :), but how would you check for existence of some particular features in particular row? For example if you use bin-search over `X_indices` it will take O(log(N)), where N is number of initialised features in current row. It takes way more time than O(1) access to value in array.

+ I don't think It's common to have such big feature space, even in sparse matrix. Implemented solution will work faster for 99% use cases.
+ Again, I tried to keep it consistent with _tree.pyx